File: 백준/Platinum/14245. XOR/XOR.py
import sys


input = lambda: sys.stdin.readline().rstrip()


def build(node, start, end):
    if start == end:
        tree[node] = arr[start]
        return

    mid = start + end >> 1
    build(node * 2, start, mid)
    build(node * 2 + 1, mid + 1, end)
    tree[node] = tree[node * 2] ^ tree[node * 2 + 1]


def propagation(node, start, end):
    if lazy[node] != 0:
        tree[node] ^= lazy[node] * (end - start + 1)

        if start != end:
            lazy[node * 2] ^= lazy[node]
            lazy[node * 2 + 1] ^= lazy[node]

        lazy[node] = 0


def update(left, right, val, node, start, end):
    propagation(node, start, end)

    if right < start or end < left:
        return

    if left <= start and end <= right:
        tree[node] ^= val * (end - start + 1)

        if start != end:
            lazy[node * 2] ^= val
            lazy[node * 2 + 1] ^= val
        return

    mid = start + end >> 1
    update(left, right, val, node * 2, start, mid)
    update(left, right, val, node * 2 + 1, mid + 1, end)
    tree[node] = tree[node * 2] ^ tree[node * 2 + 1]


def query(left, right, node, start, end):
    propagation(node, start, end)

    if right < start or end < left:
        return 0

    if left <= start and end <= right:
        return tree[node]

    mid = start + end >> 1

    return query(left, right, node * 2, start, mid) ^ query(left, right, node * 2 + 1, mid + 1, end)


if __name__ == '__main__':
    N = int(input())
    arr = list(map(int, input().split()))
    tree = [0] * (4 * N)
    lazy = [0] * (4 * N)
    build(1, 0, N - 1)
    M = int(input())

    for _ in range(M):
        q, *data = map(int, input().split())

        if q == 1:
            a, b, c = data
            update(a, b, c, 1, 0, N - 1)
        else:
            a = data[0]
            print(query(a, a, 1, 0, N - 1))

File: 백준/Platinum/12844. XOR/XOR.cc
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;

const int MAX = 2000000;

unsigned int tree[2000001];
unsigned int lazy[2000001];
int N;
vector<unsigned int> A;

void build(int node, int start, int end)
{
    if (start == end)
    {
        tree[node] = A[start];
    }
    else
    {
        int mid = (start + end) / 2;
        build(2 * node, start, mid);
        build(2 * node + 1, mid + 1, end);
        tree[node] = tree[2 * node] ^ tree[2 * node + 1];
    }
}

void push_down(int node, int start, int end)
{
    if (lazy[node] != 0)
    {
        int mid = (start + end) / 2;
        int left = 2 * node;
        int right = 2 * node + 1;

        int left_size = mid - start + 1;
        int right_size = end - mid;

        if (left_size % 2 == 1)
        {
            tree[left] ^= lazy[node];
        }
        lazy[left] ^= lazy[node];

        if (right_size % 2 == 1)
        {
            tree[right] ^= lazy[node];
        }
        lazy[right] ^= lazy[node];

        lazy[node] = 0;
    }
}

void update(int node, int start, int end, int left, int right, unsigned int val)
{
    if (right < start || end < left)
    {

        return;
    }
    if (left <= start && end <= right)
    {
        if ((end - start + 1) % 2 == 1)
        {
            tree[node] ^= val;
        }
        lazy[node] ^= val;
        return;
    }

    push_down(node, start, end);
    int mid = (start + end) / 2;
    update(2 * node, start, mid, left, right, val);
    update(2 * node + 1, mid + 1, end, left, right, val);
    tree[node] = tree[2 * node] ^ tree[2 * node + 1];
}

unsigned int query(int node, int start, int end, int left, int right)
{
    if (right < start || end < left)
    {

        return 0;
    }
    if (left <= start && end <= right)
    {
        return tree[node];
    }

    push_down(node, start, end);
    int mid = (start + end) / 2;
    unsigned int p1 = query(2 * node, start, mid, left, right);
    unsigned int p2 = query(2 * node + 1, mid + 1, end, left, right);
    return p1 ^ p2;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> N;
    A.resize(N);

    for (int i = 0; i < N; i++)
        cin >> A[i];

    memset(lazy, 0, sizeof(lazy));

    build(1, 0, N - 1);

    int M;
    cin >> M;

    string result;
    result.reserve(M * 20);

    while (M--)
    {
        int q;
        cin >> q;
        if (q == 1)
        {
            int i, j;
            unsigned int k;
            cin >> i >> j >> k;
            update(1, 0, N - 1, i, j, k);
        }
        else if (q == 2)
        {
            int i, j;
            cin >> i >> j;
            unsigned int ret = query(1, 0, N - 1, i, j);
            result += to_string(ret) + "\n";
        }
    }

    cout << result;
}